market/mysterium/mysterium_api.go

```diff
@@ -235,12 +238,28 @@ func (mApi *MysteriumAPI) QueryProposals(query ProposalsQuery) ([]market.Service
 	if err != nil {
 		return nil, err
 	}
+	req.Header.Add("If-None-Match", mApi.latestProposalsEtag)
 
-	var proposalsResponse ProposalsResponse
-	err = mApi.http.DoRequestAndParseResponse(req, &proposalsResponse)
+	res, err := mApi.http.Do(req)
 	if err != nil {
 		return nil, err
 	}
+	defer res.Body.Close()
+
+	if res.StatusCode == http.StatusNotModified {
+		return nil, errors.New("node already contains newest proposals")
+	}
+
+	if err := requests.ParseResponseError(res); err != nil {
+		return nil, err
+	}
+
+	var proposalsResponse ProposalsResponse
+	if err := requests.ParseResponseJSON(res, &proposalsResponse); err != nil {
+		return nil, fmt.Errorf("cannot parse proposals response: %v", err)
+	}
+
+	mApi.latestProposalsEtag = res.Header.Get("ETag")
```

Comment:
This `latestProposalsEtag` does not thread-safe. It can be called from different places at the same time.

Reply:
Thanks. Initially I thought that only fetcher can pull proposals but now I see that there are more places. My current solution is not going to work. I propose that I modify proposals [fetcher.go](https://github.com/mysteriumnetwork/node/blob/master/core/discovery/api/fetcher.go) and hold etag state inside this struct. @soffokl Does that make sense?

Reply:
Fetcher is using mysteriumAPI `QueryProposals` too. So it's enough to keep it only here. My suggestion is just to protect this field by the mutex to avoid unexpected panic.